### PR TITLE
Feature/protect token route

### PR DESCRIPTION
**Pull request by kpchoy**
https://issues.corp.twilio.com/browse/VBLOCKS-1460

Protect Token Route with auth0 authentication

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by mhuynh5757** [comment]
Do we have a unit test for the unhappy route in `server.test.ts`? Meaning when the middleware would send a `401`.

**Comment by kpchoy** [comment]
> Do we have a unit test for the unhappy route in `server.test.ts`? Meaning when the middleware would send a `401`.

Not yet, will work on adding

**Comment by charliesantos** [comment]
@kpchoy this PR is showing changes that are already in the public beta branch. Can you please create a separate PR for updating [feature/3rd_party_auth](https://github.com/twilio/twilio-voice-react-native-app/tree/feature/3rd_party_auth) branch, review and merge that first, then we can review this current PR. That way, we can only see what you changed.

**Comment by kpchoy** [comment]
> @kpchoy this PR is showing changes that are already in the public beta branch. Can you please create a separate PR for updating [feature/3rd_party_auth](https://github.com/twilio/twilio-voice-react-native-app/tree/feature/3rd_party_auth) branch, review and merge that first, then we can review this current PR. That way, we can only see what you changed.

Sure will do that

File: app/src/screens/SignIn/index.tsx [hunk 1]
```diff
@@ -38,15 +40,22 @@ const styles = StyleSheet.create({
...
+  const dispatch = useDispatch();
 
   const onLogin = async () => {
     try {
-      await authorize({ scope: config.auth0Scope });
+      await authorize({
+        scope: config.auth0Scope,
+        audience: config.audience,
+      });
+      const { accessToken } = await getCredentials();
```
Comment (mhuynh5757) [inline app/src/screens/SignIn/index.tsx:52]: I think this sort of cascading action logic should belong in the store/actions file. Let's think of an async action that performs this sort of authentication -> token fetching logic in one action we can dispatch from this component.

Comment (mhuynh5757) [reply app/src/screens/SignIn/index.tsx:52]: Actually, I can see the merit of two separate actions here. Let's move the authorize call into an action, though.

Comment (mhuynh5757) [reply app/src/screens/SignIn/index.tsx:52]: Does Auth0 have utility functions for this sort of functionality outside of React hooks? Could we fabricate our own actions instead?

File: app/src/store/user.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { createSlice } from '@reduxjs/toolkit';
+import type { PayloadAction } from '@reduxjs/toolkit';
+
+interface UserState {
+  accessToken: string;
+}
+
```
Comment (mhuynh5757) [inline app/src/store/user.ts:7]: Ideally, this is where I'd like all the authentication functionality to behave, i.e. calls to Auth0, getting Auth0 tokens, etc.

File: server/src/server.ts [hunk 3]
```diff
@@ -3,18 +3,25 @@ import type { TwilioCredentials } from './common/types';
 import { createTokenRoute } from './routes/token';
 import { createTwimlRoute } from './routes/twiml';
 import { createLogMiddleware } from './middlewares/log';
+import { auth } from 'express-oauth2-jwt-bearer';
+import config from '../config';
 
 export function createExpressApp(twilioCredentials: TwilioCredentials) {
```
Comment (mhuynh5757) [inline server/src/server.ts:9]: We should think of some way to incorporate the auth configs into a similar mechanism. Maybe we rename from `twilioCredentials` to `serverConfig` or something similar, but I don't think we should bake configs into more than one file or mechanism (env vars).

File: app/config.ts [hunk 4]
```diff
@@ -2,6 +2,7 @@ const config = {
   clientId: '{ClientID}',
   domain: '{DomainName}',
   auth0Scope: 'openid profile email',
+  audience: '{yourApiIdentifier}',
```
Comment (charliesantos) [inline app/config.ts:5]: What is this new config for?
We should probably add docstrings for each of this field.

File: server/config.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+const config = {
+  audience: '{yourApiIdentifier}',
+  issuerBaseURL: `https://{yourDomain}/`,
```
Comment (charliesantos) [inline server/config.ts:3]: Do we expect customers to update this file? If so, these values should go to environment variables.

File: app/src/store/voice/token.ts [hunk 6]
```diff
@@ -6,11 +6,13 @@ export const getToken = createAsyncThunk<
...
   void,
   { state: State; dispatch: Dispatch }
->('voice/getToken', async () => {
+>('voice/getToken', async (_, thunkAPI) => {
+  const accessToken = thunkAPI.getState().voice.user.accessToken;
   const res = await fetch(`${defaultUrl}/token`, {
     method: 'POST',
     headers: {
       'Content-Type': 'application/json',
+      authorization: `Bearer ${accessToken}`,
```
Comment (charliesantos) [inline app/src/store/voice/token.ts:15]:
```suggestion
      Authorization: `Bearer ${accessToken}`,
```

File: server/src/server.ts [hunk 7]
```diff
@@ -3,18 +3,25 @@ import type { TwilioCredentials } from './common/types';
...
+
   app.use(express.urlencoded({ extended: false }));
   app.use(express.json());
 
   app.use(createLogMiddleware());
 
   const tokenRouter = Router();
   tokenRouter.use(createTokenRoute(twilioCredentials));
-  app.post('/token', tokenRouter);
+  app.post('/token', jwtCheck, tokenRouter);
```
Comment (charliesantos) [inline server/src/server.ts:24]: Oh this is cool! This means any endpoint that we want to protect, we just do this same thing?

**Review by charliesantos — CHANGES_REQUESTED** [review]

File: app/config.ts (hunk 4 above)
Comment (kpchoy) [reply app/config.ts:5]: What kind of docstring do you have in mind? Reference link to auth0?

File: server/src/server.ts (hunk 7 above)
Comment (kpchoy) [reply server/src/server.ts:24]: Yup we would just need to add `jwtcheck` to each route we want to protect

File: app/src/screens/SignIn/index.tsx (hunk 1 above)
Comment (mhuynh5757) [reply app/src/screens/SignIn/index.tsx:52]: That's a good point. My opinion is that this kind of logic seems very "atomic" in the sense we don't really need to use the steps individually anywhere in the app. The hook they provide seems more for apps that don't use centralized logic, but I can see your point.

File: app/config.ts (hunk 4 above)
Comment (charliesantos) [reply app/config.ts:5]: Something like the following. I added an example for clientId. Just fill in the docstrings for the rest.
```ts
/**
 * Auth0 related configurations
 */
const config = {
  /**
   * The unique identifier for your Auth0 application.
   * See https://auth0.com/docs/get-started/applications/application-settings#basic-settings
   */
  clientId: '{ClientID}',
  domain: '{DomainName}',
  auth0Scope: 'openid profile email',
  audience: '{yourApiIdentifier}',
}
```

Comment (charliesantos) [reply app/config.ts:5]: I understand we don't do this in most part of the code. But I think we should start doing it on areas where things are not self explanatory. For example on this new config, I'm not sure what audience really means :D

**Review by mhuynh5757 — COMMENTED** [review]
The server config stuff is looking good!

File: server/src/server.ts (hunk 7 above)
Comment (charliesantos) [reply server/src/server.ts:23]: Not for this PR, but is it possible to get other user information such as email or username from the jwt? Or can you post an example jwt?

**Review by charliesantos — APPROVED** [review]
Looks really good. +1 for me. Let's wait for @mhuynh5757 's +1 before merging.

File: server/src/server.ts (hunk 7 above)
Comment (charliesantos) [reply server/src/server.ts:23]: Ok, we can take a look at that later. But can you please post an example jwt here? I like to see it's structure.

**Pull request merged**